e2e/cluster/cmx/cluster.go

```diff
@@ -0,0 +1,463 @@
+		if err != nil {
+			return fmt.Errorf("failed to get SSH endpoint for node %d: %v", i, err)
+		}
+		c.Nodes[i].sshEndpoint = sshEndpoint
+	}
+	return nil
+}
+
+func (c *Cluster) waitUntilAirgapped(node int) error {
+	timeout := time.After(2 * time.Minute)
+	tick := time.Tick(5 * time.Second)
+
+	for {
+		select {
+		case <-timeout:
+			return fmt.Errorf("timed out waiting for node to be airgapped after 1 minute")
+		case <-tick:
+			_, _, err := c.RunCommandOnNode(node, []string{"curl", "--connect-timeout", "5", "www.google.com"})
+			if err != nil {
+				c.t.Logf("node %d is airgapped successfully", node)
+				return nil
+			}
+			c.t.Logf("node %d is not airgapped yet", node)
+		}
+	}
+}
+
+func (c *Cluster) Cleanup(envs ...map[string]string) {
+	c.generateSupportBundle(envs...)
+	c.copyPlaywrightReport()
+	c.Destroy()
+}
+
+func (c *Cluster) Destroy() {
+	for _, node := range c.Nodes {
+		cmd := exec.Command("replicated", "vm", "rm", node.ID)
+		output, err := cmd.CombinedOutput()
+		if err != nil {
+			c.t.Logf("failed to destroy node %s: %v: %s", node.Name, err, string(output))
+		}
+	}
+}
+
+func (c *Cluster) RunCommandOnNode(node int, line []string, envs ...map[string]string) (string, string, error) {
+	return runCommandOnNode(c.Nodes[node], line, envs...)
+}
+
+func runCommandOnNode(node Node, line []string, envs ...map[string]string) (string, string, error) {
+	for _, env := range envs {
+		for k, v := range env {
+			line = append([]string{fmt.Sprintf("%s=%s", k, v)}, line...)
+		}
+	}
+	line = append([]string{"sudo"}, line...)
+
+	cmd := exec.Command("ssh", append(sshArgs(), node.sshEndpoint, strings.Join(line, " "))...)
+	var stdout, stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+
+	if exitErr, ok := err.(*exec.ExitError); ok && exitErr.ExitCode() == 255 {
+		// check if this is a reset-installation command that resulted in exit code 255
+		// as this is expected behavior when the node reboots and the ssh connection is lost
+		if strings.Contains(strings.Join(line, " "), "reset-installation") {
+			return stdout.String(), stderr.String(), nil
+		}
+	}
+
+	return stdout.String(), stderr.String(), err
+}
+
+func (c *Cluster) SetupPlaywrightAndRunTest(testName string, args ...string) (string, string, error) {
+	if err := c.SetupPlaywright(); err != nil {
+		return "", "", fmt.Errorf("failed to setup playwright: %w", err)
+	}
+	return c.RunPlaywrightTest(testName, args...)
+}
+
+func (c *Cluster) SetupPlaywright(envs ...map[string]string) error {
+	c.t.Logf("%s: bypassing kurl-proxy", time.Now().Format(time.RFC3339))
+	_, stderr, err := c.RunCommandOnNode(0, []string{"bypass-kurl-proxy.sh"}, envs...)
+	if err != nil {
+		return fmt.Errorf("fail to bypass kurl-proxy: %v: %s", err, string(stderr))
+	}
+	c.t.Logf("%s: installing playwright", time.Now().Format(time.RFC3339))
+	cmd := exec.Command("sh", "-c", "cd playwright && npm ci && npx playwright install --with-deps")
```

Comment:
is this quick each subsequent time it is run? i assume it is run on the runner once for each test. should we just make this a required dependency like the replicated cli?